optuna/storages/_journal/file.py

```diff
@@ -0,0 +1,184 @@
+                return True
+            except OSError as err:
+                if err.errno == errno.EEXIST:
+                    time.sleep(sleep_secs)
+                    sleep_secs = min(sleep_secs * 2, 1)
+                    continue
+                else:
+                    raise err
+            except BaseException:
+                self.release()
+                raise
+
+    def release(self) -> None:
+        try:
+            os.rename(self._lockfile, self._lockrenamefile)
+            os.unlink(self._lockrenamefile)
+        except OSError:
+            raise RuntimeError("Error: did not possess lock")
+        except BaseException:
+            os.unlink(self._lockrenamefile)
+            raise
+
+
+@contextmanager
+def get_lock_file(lock_obj: JournalFileBaseLock) -> Iterator[None]:
+    lock_obj.acquire()
+    try:
+        yield
+    finally:
+        lock_obj.release()
+
+
+class JournalFileStorage(BaseJournalLogStorage):
+    """File storage class for Journal log backend.
+
+    Args:
+        file_path:
+            Path of file to persist the log to.
+
+        lock_obj:
+            Lock object for process exclusivity.
+
+    """
+
+    def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = None) -> None:
+        self._file_path: str = file_path
+        self._lock = lock_obj or JournalFileSymlinkLock(self._file_path)
+        open(self._file_path, "a").close()  # Create a file if it does not exist
+        self._log_number_offset: Dict[int, int] = {0: 0}
+
+    def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
+        logs = []
+        with open(self._file_path, "r") as f:
+            log_number_start = 0
+            if log_number_from in self._log_number_offset:
+                f.seek(self._log_number_offset[log_number_from])
+                log_number_start = log_number_from
+
+            last_decode_error = None
+            for log_number, line in enumerate(f, start=log_number_start):
+                if last_decode_error is not None:
+                    raise last_decode_error
+                if log_number + 1 not in self._log_number_offset:
+                    byte_len = len(line.encode("utf-8"))
+                    self._log_number_offset[log_number + 1] = (
+                        self._log_number_offset[log_number] + byte_len
+                    )
+                if log_number < log_number_from:
+                    continue
+                try:
+                    logs.append(json.loads(line))
+                except json.JSONDecodeError as err:
+                    last_decode_error = err
+                    del self._log_number_offset[log_number + 1]
+
+            return logs
+
+    def append_logs(self, logs: List[Dict[str, Any]]) -> None:
+        with get_lock_file(self._lock):
+            what_to_write = ""
+            for log in logs:
+                what_to_write += json.dumps(log) + "\n"
```

Comment:
It looks that we can simply use `"\n".join()` here.

```suggestion
            what_to_write = "\n".join([json.dumps(log) for log in logs])
```

Reply:
https://github.com/optuna/optuna/pull/3854/commits/bcce9487ecfc1a6c14b187653fe523afe3006e4b